packages/phalanx/src/Phalanx_Field_Def.hpp

```diff
@@ -0,0 +1,297 @@
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Roger Pawlowski (rppawlo@sandia.gov), Sandia
+// National Laboratories.
+//
+// ************************************************************************
+// @HEADER
+
+
+#ifndef PHX_FIELD_DEF_HPP
+#define PHX_FIELD_DEF_HPP
+
+#include <algorithm>
+#include <sstream>
+#include <vector>
+#include "Teuchos_Assert.hpp"
+#include "Teuchos_TypeNameTraits.hpp"
+#include "Kokkos_DynRankView_Fad.hpp" // for copy/assignment specializations
+
+// **********************************************************************
+#ifdef PHX_DEBUG
+template<typename DataT,int Rank>
+const std::string PHX::Field<DataT,Rank>::m_field_tag_error_msg =
+  "Error - PHX::Field - No tag has been set!";
+
+template<typename DataT,int Rank>
+const std::string PHX::Field<DataT,Rank>::m_field_data_error_msg =
+  "Error - PHX::Field - No data has been set!  Please bind memory (call getFieldData()) to MDField!";
+#endif
+
+// **********************************************************************
+template<typename DataT,int Rank>
+PHX::Field<DataT,Rank>::
+Field(const std::string& name, const Teuchos::RCP<const PHX::DataLayout>& t) :
+  m_tag(name,t)
+#ifdef PHX_DEBUG
+  , m_tag_set(true),
+  m_data_set(false)
+#endif
+{ }
+
+// **********************************************************************
+template<typename DataT,int Rank>
+PHX::Field<DataT,Rank>::Field(const PHX::Tag<DataT>& v) :
+  m_tag(v)
+#ifdef PHX_DEBUG
+  ,m_tag_set(true),
+  m_data_set(false)
+#endif
+{ }
+
+// **********************************************************************
+template<typename DataT,int Rank>
+PHX::Field<DataT,Rank>::Field() :
+  m_tag("???", Teuchos::null)
+#ifdef PHX_DEBUG
+  ,m_tag_set(false),
+  m_data_set(false)
+#endif
+{ }
+
+// **********************************************************************
+template<typename DataT,int Rank>
+template<typename CopyDataT>
+PHX::Field<DataT,Rank>::Field(const Field<CopyDataT,Rank>& source) :
+  m_tag(source.m_tag),
+  m_field_data(source.m_field_data)
+#ifdef PHX_DEBUG
+  ,m_tag_set(source.m_tag_set),
+  m_data_set(source.m_data_set)
+#endif
+{
+  static_assert(std::is_same<typename std::decay<DataT>::type, typename std::decay<CopyDataT>::type>::value,
+                "ERROR: Compiletime MDField copy ctor requires scalar types to be the same!");
+}
+
+// **********************************************************************
+template<typename DataT,int Rank>
+PHX::Field<DataT,Rank>::~Field()
```

Comment:
no. It was a cut-n-paste from the original MDField that was written pre-c++11. I can change that over in both classes.